src/Tools/BuildValidator/CompilationDiff.cs

```diff
@@ -26,44 +37,163 @@ internal class CompilationDiff
 
-                var newBytes = peStream.ToArray();
+                        // TODO: output source files should include the entire relative path instead of just the file name.
+                        foreach (var tree in producedCompilation.SyntaxTrees)
+                        {
+                            var sourceFilePath = Path.Combine(sourcesPath, Path.GetFileName(tree.FilePath));
+                            using var file = File.OpenWrite(sourceFilePath);
+                            var writer = new StreamWriter(file);
+                            tree.GetText().Write(writer);
+                            writer.Flush();
+                        }
 
-                return new CompilationDiff(assemblyFile.FullName, newBytes.SequenceEqual(originalBytes));
+                        var originalAssemblyPath = Path.Combine(originalPath, originalBinaryPath.Name);
+                        File.WriteAllBytes(originalAssemblyPath, originalBytes);
+
+                        var rebuildAssemblyPath = Path.Combine(rebuildPath, originalBinaryPath.Name);
+                        File.WriteAllBytes(rebuildAssemblyPath, rebuildBytes);
+
+                        var originalPeMdvPath = Path.Combine(originalPath, assemblyName + ".pe.mdv");
+                        var originalPdbMdvPath = Path.Combine(originalPath, assemblyName + ".pdb.mdv");
+                        writeVisualization(originalPeMdvPath, optionsReader.PeReader.GetMetadataReader());
+                        writeVisualization(originalPdbMdvPath, optionsReader.PdbReader);
+
+                        var rebuildPeMdvPath = Path.Combine(rebuildPath, assemblyName + ".pe.mdv");
+                        var rebuildPdbMdvPath = Path.Combine(rebuildPath, assemblyName + ".pdb.mdv");
+                        fixed (byte* ptr = rebuildBytes)
+                        {
+                            var rebuildPeReader = new PEReader(ptr, rebuildBytes.Length);
+                            writeVisualization(rebuildPeMdvPath, rebuildPeReader.GetMetadataReader());
+
+                            if (rebuildPeReader.TryOpenAssociatedPortablePdb(
+                                rebuildAssemblyPath,
+                                path => File.Exists(path) ? File.OpenRead(path) : null,
+                                out var provider,
+                                out _) && provider is { })
+                            {
+                                var rebuildPdbReader = provider.GetMetadataReader(MetadataReaderOptions.Default);
+                                writeVisualization(rebuildPdbMdvPath, rebuildPdbReader);
+                            }
+                        }
+
+                        var ildasmOriginalOutputPath = Path.Combine(originalPath, assemblyName + ".il");
+                        var ildasmRebuildOutputPath = Path.Combine(rebuildPath, assemblyName + ".il");
+
+                        // TODO: can we bundle ildasm in with the utility?
```

Comment:
Good question...

Reply:
I think we can, it's just work 😉 since ilasm for all platforms is bundled in the compiler emit tests.

Reply:
Yeah this is just "annoying work we need to do" :smile:

Reply:
😅